docs/architecture/adr-044-lite-client-with-weak-subjectivity.md

```diff
@@ -121,6 +124,176 @@ network usage.
 Check out the formal specification
 [here](https://github.com/tendermint/tendermint/blob/master/docs/spec/consensus/light-client.md).
 
+### Implementation
+
+There are two primary modes of usage right now:
+
+1) Trusted RPC proxy (wrapping an RPC client + verification)
+2) Part of the IBC light client (only verification bit, no RPC) [spec](https://github.com/cosmos/ics/tree/master/spec/ics-002-client-semantics)
+
+First, we'll need something, which will provide us secure headers & validator sets.
+
+```go
+type Provider interface {
+	// 0 - latest
+	SignedHeader(height int64) (SignedHeader, error)
+
+  // 0 - latest
+  Validators(height int64) (ValidatorSet, error)
+}
+```
+
+In case of the proxy it will be a `http` provider (wrapping RPC client). For
+IBC, it will be a `ibc` provider, receiving information from IBC transactions.
+
+Once we have the information, we need to verify it.
+
+```go
+type mode int
+
+const (
+	sequential mode = iota
+	bisecting
+)
+
+// default mode - DefaultBisectingVerification
+type Verifier struct {
+	chainID            string
+	options            TrustOptions
+	lastVerifiedHeight int64
+	logger             log.Logger
+
+	mode       mode
+	trustLevel float
+
+	// Source of new headers.
+	source Provider
+
+	// Alternative sources for checking the primary for misbehavior by comparing data.
+	// If the primary misbehaves, we report the evidence to them.
+	verifiers []ProviderAndEvidenceReporter
+
+	// Where trusted headers are stored.
```

Comment:
Maybe better to say where verified headers are stored as trust is temporal property. Note that after trusted period of a header expires, we still want to keep it at least for some time so we can give time to failure detection module to find conflicting header in case of attack (fork).